charmcraft/config.py

```diff
@@ -14,171 +14,282 @@
 
-    dirpath = attr.ib(default=None)
-    config_provided = attr.ib(default=None)
-    started_at = attr.ib(default=None)
+class Parts(pydantic.BaseModel, frozen=True, validate_all=True):
+    """Definition of parts to build."""
 
+    __root__: Dict[pydantic.StrictStr, Part] = {}
 
-@attr.s(kw_only=True, frozen=True)
-class Config:
-    """Root of all the configuration."""
+    def get(self, part_name) -> Part:
+        """Get part by name.
 
-    charmhub = attr.ib(default={}, converter=CharmhubConfig.from_dict)
-    parts = attr.ib(default={}, converter=BasicPrime.from_dict)
-    type = attr.ib(default=None)
+        :returns: Part if exists, None if not.
+        """
+        try:
+            return self.__root__[part_name]
+        except KeyError:
+            return None
 
-    # this item is provided by the code itself, not the user, as convenience for the
-    # rest of the code
-    project = attr.ib(default=None)
 
+class CharmhubConfig(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Definition of Charmhub endpoint configuration."""
 
-CONFIG_SCHEMA = {
-    "type": "object",
-    "properties": {
-        "type": {"type": "string", "enum": ["charm", "bundle"]},
-        "charmhub": {
-            "type": "object",
-            "properties": {
-                "api_url": {"type": "string", "format": "url"},
-                "storage_url": {"type": "string", "format": "url"},
-            },
-            "additionalProperties": False,
-        },
-        "parts": {
-            "type": "object",
-            "properties": {
-                "bundle": {
-                    "type": "object",
-                    "properties": {
-                        "prime": {
-                            "type": "array",
-                            "items": {
-                                "type": "string",
-                                "format": "relative_path",
-                            },
-                        },
-                    },
-                },
-            },
-        },
-    },
-    "required": ["type"],
-    "additionalProperties": False,
-}
+    api_url: pydantic.HttpUrl = "https://api.charmhub.io"
+    storage_url: pydantic.HttpUrl = "https://storage.snapcraftcontent.com"
+
+
+class Project(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Internal-only project configuration."""
+
+    dirpath: pydantic.DirectoryPath
+    content: Dict[str, Any] = {}
+    config_provided: bool = False
+
+    # this timestamp will be used in several places, even sent to Charmhub: needs to be UTC
+    started_at: datetime.datetime = datetime.datetime.utcnow()
+
+
+class Config(
+    pydantic.BaseModel,
+    extra=pydantic.Extra.forbid,
+    frozen=True,
+    validate_all=True,
+):
+    """Definition of charmcraft.yaml configuration."""
+
+    type: pydantic.StrictStr
```

Comment:
Can we put something here like `StrictStr | None` or `Optional(StrictStr)`, I totally hate this new `undefined`.

Reply:
I hate it too but it was a bit painful to deal with the model being optionally None but only in certain cases. I figured since it's only short-lived, who cares.  I worked around it though.